### PR TITLE
V3

### PR DESCRIPTION
**Pull request by ballasn**
Add conv3d

### CODE REVIEW

**Comment by carriepl** [comment]
I'm currently reviewing but it seems this PR will need a rebase.

File: theano/sandbox/cuda/dnn.py [hunk 1]
```diff
@@ -127,6 +127,30 @@ def c_set_tensor4d(var, desc, err, fail):
 
         """ % dict(var=var, err=err, desc=desc, fail=fail)
 
+def c_set_tensorNd(var, nb_dim, desc, err, fail):
+    return """CudaNdarra
```
Comment (carriepl) [inline theano/sandbox/cuda/dnn.py]: ?

File: theano/sandbox/cuda/dnn.py [hunk 2]
```diff
@@ -355,10 +379,146 @@ def c_code_cache_version(self):
...
+  }
+}
+""" % dict(name=name, img_shape=img_shape, kern_shape=kern_shape, desc=desc,
+           bmode=bmode, conv_flag=conv_flag, fail=sub['fail'],
+           subsd=self.subsample[0], subsx=self.subsample[1], subsy=self.subsample[2],
+           pad_d_spec=pad_d_spec, pad_h_spec=pad_h_spec, pad_w_spec=pad_w_spec)
+
+    def c_code_cache_version(self):
+        return None
+#        return (2, version())
```
Comment (carriepl) [inline theano/sandbox/cuda/dnn.py]: Looks like you meant to restore this line before commiting

File: theano/sandbox/cuda/dnn.py [hunk 3]
```diff
@@ -355,10 +379,146 @@ def c_code_cache_version(self):
...
+
+
+
```
Comment (carriepl) [inline theano/sandbox/cuda/dnn.py]: Only 2 empty lines are required

File: theano/sandbox/cuda/dnn.py [hunk 4]
```diff
@@ -526,6 +701,101 @@ def infer_shape(self, node, shape):
...
+class GpuDnnConv3d(GpuDnnConv):
+    """
+    The forward convolution.
+
+    :param image:
+    :param kernel:
+    :param descr: the convolution descriptor
+    """
+    __props__ = ('workmem', 'inplace')
+    __input_name__ = ('image', 'kernel', 'output',
```
Comment (carriepl) [inline theano/sandbox/cuda/dnn.py]: __input_name__ needs to be adjusted to refect the new input you added. This also goes for the other classes.

File: theano/sandbox/cuda/dnn.py [hunk 5]
```diff
@@ -526,6 +701,101 @@ def infer_shape(self, node, shape):
...
+        if output.type.ndim != 5:
+            raise TypeError('output must be a 5D tensor')
+        if not isinstance(desc.type, CDataType) \
+                or desc.type.ctype != 'cudnnConvolutionDescriptor_t':
+            raise TypeError('desc must be cudnnConvolutionDescriptor_t')
+        alpha = ensure_float(alpha, _one, 'alpha')
+        beta = ensure_float(beta, _zero, 'beta')
+        nb_dim = ensure_int(nb_dim, _ifive, 'nb_dim')
+
+
```
Comment (carriepl) [inline theano/sandbox/cuda/dnn.py]: remove this line

File: theano/sandbox/cuda/dnn.py [hunk 6]
```diff
@@ -617,13 +888,70 @@ def make_node(self, img, topgrad, output, desc, alpha=None, beta=None):
...
+    :param image:
+    :param kernel:
+    :param descr: the convolution descriptor
+
+    """
+    __props__ = ('workmem', 'inplace',)
+    __input_name__ = ('image', 'grad', 'output', 'descriptor', 'alpha', 'beta')
+
+    def __init__(self, inplace=False, workmem=None):
```
Comment (carriepl) [inline theano/sandbox/cuda/dnn.py]: The workmem argument is useless in this class. Some (but not all) of your other 3d classes also have this

File: theano/sandbox/cuda/dnn.py [hunk 7]
```diff
@@ -778,6 +1165,95 @@ def dnn_conv(img, kerns, border_mode='valid', subsample=(1, 1),
...
+        If border_mode is 'valid', subsample is (1,1,1) and direction_hint is
+        'bprop weights', it will use GpuDnnConvGradW.
+        If border_mode is 'full', subsample is (1,1,1) and direction_hint is
+        *not* 'forward!', it will use GpuDnnConvGradI.
+        This parameter is used internally by graph optimizers and may be
+        removed at any time without a deprecation period. You have been warned.
+    :param workmem: Specify the amount of working memory allowed.
+      More memory is usually faster.  One of 'none', 'small' or
+      'large'.  (default is None which takes its value from
+      :attr:`config.dnn.conv.workmem`)
```
Comment (carriepl) [inline theano/sandbox/cuda/dnn.py]: This conflicts with another comment above that says that only the 'none' implementation works.

**Comment by ballasn** [comment]
updated,

**Comment by carriepl** [comment]
Seems mostly good. For additional coverage (and for future proofing, why not?), you should add some tests for cases analogous to those that cause issue with some implementations of conv2D such as : 1x1 filter, inputs or kernels with dimensions larger than 1024, inputs with dimensions larger than an int16 could hold, ...

**Comment by carriepl** [comment]
After IRL discussion, we agreed to merge now and to add the tests in a follow up PR.

**Pull request merged**